crux-lucene/src/crux/lucene.clj

```diff
@@ -47,17 +47,21 @@
             (.add (StoredField. "latest-completed-tx" ^long (:crux.tx/tx-id tx))))]
     (.updateDocument index-writer t d)))
 
+(defn- ->index-searcher [lucene-store]
+  (let [^SearcherManager searcher-manager (:searcher-manager lucene-store)
+        s (.acquire searcher-manager)]
+    [s (fn [] (.release searcher-manager s))]))
+
 (defn latest-submitted-tx [lucene-store]
-  (let [{:keys [^Directory directory]} lucene-store]
-    (when (DirectoryReader/indexExists directory)
-      (with-open [directory-reader (DirectoryReader/open directory)]
-        (let [index-searcher (IndexSearcher. directory-reader)
-              q (TermQuery. (Term. "meta" "latest-completed-tx"))
-              d ^ScoreDoc (first (.-scoreDocs (.search index-searcher q 1)))]
-          (when d
+  (let [[^IndexSearcher index-searcher index-searcher-release-fn] (->index-searcher lucene-store)
```

Comment:
suspect you want some kind of try-finally here so that the release-fn is called whatever happens in the rest of this function?

Reply:
maybe you want the return from `->index-searcher` to implement `Closeable` so that you can use `with-open`?

Reply:
I agree returning something `Closeable` would be nicer but I wasn't certain if it was worth the effort to figure out. I couldn't see a way to wrap this other usage instance inside `cio/->cursor` in a `with-open`: https://github.com/juxt/crux/pull/1468/files#diff-97fb172b1e6d0b14e8f78e6bdcb364d8fef607cf1ad68e42d97e6404744b9e9fR93-R94

Would you call `(.close index-searcher)` there explicitly instead? Or is there some better way to tie the two `Closeable` things together?

> suspect you want some kind of try-finally here so that the release-fn is called whatever happens in the rest of this function?
I'm not sure if anything in the rest of the function is ever going to throw (at least without having much more severe consequences) - are you thinking to avoid possibilities of memory leaks?

Reply:
> I'm not sure if anything in the rest of the function is ever going to throw

Assume it might :smile: We should always take responsibility for closing resources that we own.

> are you thinking to avoid possibilities of memory leaks?

Without knowing the specific implementation details of `IndexSearcher`, I don't _know_ what the impact of not releasing it might be. In Rocks, though, for example, not closing a snapshot has the effect of preventing Rocks from compacting any transactions older than that snapshot - i.e. the impact of _not_ closing something can be large.

Reply:
In this case, because you're handing off a cursor to your caller, you don't actually want to close the resource in a `finally`, because you're not done with it when the function returns.

```clojure
(defn search [lucene-store, ^Query q]
  (assert lucene-store)
  (let [[^IndexSearcher index-searcher index-searcher-release-fn] (->index-searcher lucene-store)]
    (try
      (let [q (FunctionScoreQuery. q (DoubleValuesSource/fromQuery q))
            score-docs ...]

        (when (seq score-docs)
          (log/debug (.explain index-searcher q (.-doc ^ScoreDoc (first score-docs)))))

        (cio/->cursor (fn []
                        (index-searcher-release-fn))
                      (->> score-docs
                           (map (fn [^ScoreDoc d]
                                  (vector (.doc index-searcher (.-doc d))
                                          (.-score d)))))))
      (catch Exception e
        (index-searcher-release-fn)
        (throw e)))))
```

Reply:
I've now added the try+catch, thanks for the feedback!

For posterity: we also discussed the merits of creating a `IndexSearcherWrapper` record that implements `closeable`...but it's probably overkill :)